d3-chord/index.d.ts

```diff
@@ -1,78 +1,401 @@
+     * target object returned by the source or target accessor, respectively.
+     */
+    endAngle(): (this: This, d: RibbonSubgroupDatum, ...args: any[]) => number;
+    /**
+     * Sets the end angle to a fixed number in radians and returns this ribbon generator.
+     *
+     * @param angle A fixed numeric value for the end angle in radians.
+     */
     endAngle(angle: number): this;
-    endAngle(angle: (this: This, d: ChordSubgroupDatum, ...args: any[]) => number): this;
+    /**
+     * Sets the end angle accessor to the specified function and returns this ribbon generator.
+     *
+     * @param angle An accessor function which is invoked for the source and target of the chord. The accessor function is invoked in the same "this" context as the generator was invoked in and
+     * receives as the first argument the source or target object returned by the respective source or target accessor function of the generator. It is also passed any additional arguments that were passed
+     * into the generator, with the exception of the first element representing the chord datum itself. The function returns the end angle in radians.
+     */
+    endAngle(angle: (this: This, d: RibbonSubgroupDatum, ...args: any[]) => number): this;
+
+    /**
+     * Returns the current rendering context, which defaults to null.
+     */
     context(): CanvasRenderingContext2D | null;
+    /**
+     * Sets the rendering context and returns this ribbon generator.
+     *
+     * If the context is not null, then the generated ribbon is rendered to this context as a sequence of path method calls.
+     *
+     * @param context The rendering context.
+     */
     context(context: CanvasRenderingContext2D): this;
+    /**
+     * Sets the rendering context to null and returns this ribbon generator.
+     *
+     * A path data string representing the generated ribbon will be returned when the generator is invoked with data. See also d3-path.
+     *
+     * @param context null, to remove rendering context.
+     */
     context(context: null): this;
 }
 
-export function ribbon(): RibbonGenerator<any, Chord, ChordSubgroup>;
+/**
+ * Creates a new ribbon generator with the default settings.
+ */
+export function ribbon(): RibbonGenerator<any, Ribbon, RibbonSubgroup>;
+/**
+ * Creates a new ribbon generator with the default settings.
+ *
+ * Accessor functions must be configured for the ribbon generator, should the datum types differ from the defaults.
+ *
+ * The first generic corresponds to the datum type representing a chord for which the ribbon is to be generated. The default type is Chord.
+ *
+ * The second generic corresponds to the datum type of the chord subgroup, i.e. source or target of the cord. The default type is ChordSubgroup.
+ */
 export function ribbon<Datum, SubgroupDatum>(): RibbonGenerator<any, Datum, SubgroupDatum>;
```

Comment:
Good question you are raising here. The `any` vs `{}` for the `this` context is actually referring to a general pattern in use right now for all those generators which can render to `canvas` or generate a path string, e.g. when used on a selection of `SVGPathElement`.

So the question generalizes to e.g. to `arc()` in d3-shape etc. With that in mind:
* You are right that there is a risk that someone uses one of the factory overloads that default the `this` context to `any` and then inadvertently invokes a method on `this` in a callback without further thought to the actual nature of the invocation context. And yes, there is some limited protection afforded when using `{}`
* The reason I note the _limited_ is that if someone creates a variable/constant to hold a callback function and sets a differing `this` context, e.g. for re-usability or code readability,
```ts
let cb: (this: SVGPathElement, d: ArcDatum) => number = function(this: SVGPathElement, d: ArcDatum) {
    let b: string | null = this.baseURI;
    return 10;
};
```
then the `{}` no longer holds. I.e.
```ts
let svgArc = d3Shape.arc<ArcDatum>(); // of type Arc<{}, ArcDatum> with the changed default

svgArc.cornerRadius(cb); // does NOT fail!
```
Although, this would fail:
```ts
let svgArc = d3Shape.arc<ArcDatum>(); // of type Arc<{}, ArcDatum> with the changed default

svgArc.cornerRadius(function(d) {
   let b: string | null = this.baseUri; // fails as baseUri is not defined on '{}'
   return 10;
});
```
* There is another implication: The "shorthands" factory signatures where `any` is used were intended as convenience functions,in cases, where callbacks do not utilize `this` for the use case at hand. Obviously, you are right there is not guarantee that someone does not try. Changing `any` to `{}` does not directly affect the convenience of using the generator in a typical `svg` path setting, where the `this` context is not explicitly required. I.e. for the `svgArc` from above:
```ts
let pArc: Selection<SVGPathElement, ArcDatum, any, any> = select<SVGPathElement, ArcDatum>('.arc-paths'); // mock

// configure the accessors to work with ArcDatum only

pArc.attr('d', svgArc); // compiles!
```
However, trying the same when rendering to a `canvas` rendering context, fails under normal circumstances, e.g.:
```ts
let canvasArc = d3Shape.arc();

// configure canvasArc including assigning the canvas rendering context
// Have arcDefaultDatum defined

canvasArc(arcDefaultDatum); // fails to compile, as 'this' context of '{}' cannot be assigned to 'this' of type 'void'
```
The only way to get it to work with a `canvas` setting would be to use the most extensive factory signature with `This` generic set to `void`, so that one has `Arc<void, ...>`

So with all that being said, `any` seemed somewhat easier, while not perfect.